File: Java/boj/Intermediate/boj_11438.java
package boj.Intermediate;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class boj_11438 {
    static final int LOG = 21;
    static int N, M;
    static List<Integer>[] tree;
    static int[][] parents;
    static boolean[] visited;
    static int[] depth;

    private static int lca(int a, int b) {
        if (depth[a] > depth[b]) {
            int tmp = b;
            b = a;
            a = tmp;
        }

        for (int i = LOG - 1; i >= 0; i--)
            if (depth[b] - depth[a] >= (1 << i))
                b = parents[b][i];

        if (a == b) return a;

        for (int i = LOG - 1; i >= 0; i--) {
            if (parents[a][i] != parents[b][i]) {
                a = parents[a][i];
                b = parents[b][i];
            }
        }

        return parents[a][0];
    }

    private static void setParents() {
        dfs(1, 0);

        for (int i = 1; i < LOG; i++)
            for (int j = 1; j <= N; j++)
                parents[j][i] = parents[parents[j][i - 1]][i - 1];
    }

    private static void dfs(int node, int dth) {
        visited[node] = true;
        depth[node] = dth;

        for (int next : tree[node]) {
            if (!visited[next]) {
                parents[next][0] = node;
                dfs(next, dth + 1);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());

        visited = new boolean[N + 1];
        depth = new int[N + 1];
        parents = new int[N + 1][LOG];
        tree = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) tree[i] = new ArrayList<>();

        for (int i = 1; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            tree[a].add(b);
            tree[b].add(a);
        }

        setParents();
        M = Integer.parseInt(br.readLine());

        for (int i = 0; i < M; i++) {
            String[] arg = br.readLine().split("\\s");
            sb.append(lca(Integer.parseInt(arg[0]), Integer.parseInt(arg[1])));
            sb.append('\n');
        }

        System.out.print(sb);
        br.close();
    }
}

File: Kotlin/src/main/kotlin/boj/intermediate/boj_11438.kt
package boj.intermediate

import java.lang.StringBuilder

private const val LOG = 18
private var n = 0
private var m = 0
private lateinit var tree: Array<ArrayList<Int>>
private lateinit var parents: Array<IntArray>
private lateinit var depth: IntArray
private lateinit var visited: BooleanArray

private fun lca(a: Int, b: Int): Int {
    var va = a
    var vb = b

    if (depth[va] > depth[vb]) va = vb.also { vb = va }

    for (i in LOG - 1 downTo 0)
        if (depth[vb] - depth[va] >= (1 shl i))
            vb = parents[vb][i]

    if (va == vb) return va

    for (i in LOG - 1 downTo 0) {
        if (parents[va][i] != parents[vb][i]) {
            va = parents[va][i]
            vb = parents[vb][i]
        }
    }

    return parents[va][0]
}

private fun setParents() {
    fun dfs(node: Int, dth: Int) {
        visited[node] = true
        depth[node] = dth

        for (next in tree[node]) {
            if (!visited[next]) {
                parents[next][0] = node
                dfs(next, dth + 1)
            }
        }
    }

    dfs(1, 0)
    for (i in 1 until LOG)
        for (j in 1..n)
            parents[j][i] = parents[parents[j][i - 1]][i - 1]
}

fun boj_11438() = with(System.`in`.bufferedReader()) {
    val sb = StringBuilder()

    n = readLine().toInt()
    tree = Array(n + 1) { ArrayList() }
    parents = Array(n + 1) { IntArray(LOG) }
    depth = IntArray(n + 1)
    visited = BooleanArray(n + 1)

    repeat(n - 1) {
        val (a, b) = readLine().split(" ").map { it.toInt() }
        tree[a].add(b)
        tree[b].add(a)
    }

    setParents()
    m = readLine().toInt()

    repeat(m) {
        val (a, b) = readLine().split(" ").map { it.toInt() }
        sb.append(lca(a, b)).append('\n')
    }

    print(sb)
    close()
}